### PR TITLE
Better linux build support

### PR DESCRIPTION
**Pull request by biast12**
Improves the bat files
Added bash/.sh files so linux can easily create the env and build too
Added support for building with other architectures too

### CODE REVIEW

**Comment by biast12** [comment]
This wiki page might also need to be updated to include the .sh files, but i don't think i'm able to update that https://github.com/DevilXD/TwitchDropsMiner/wiki/Setting-up-the-environment,-building-and-running

File: build.spec
Comment (guihkx) [inline build.spec:56]: Any strong reason for the switch to Ayatana's AppIndicator here?

File: .github/workflows/ci.yml
Comment (guihkx) [inline .github/workflows/ci.yml:133]: If the intention is to move away from the regular AppIndicator implementation to Ayatana's, you can drop the `gir1.2-appindicator3-0.1` package here.

File: .github/workflows/ci.yml
Comment (guihkx) [inline .github/workflows/ci.yml:201]: Same as above. Also, please note that if you intend to move to Ayatana's AppIndicator implementation, you'll have to also update the AppImage build recipe (`appimage/AppImageBuilder.yml`) to include all dependencies required by Aytana.

File: build.spec
Comment (biast12) [reply build.spec:56]: I couldn't get libappindicator3 to work on my Raspberry pi 5 (Raspberry os), and from the research i could find about it does it seem like an issue overall that it, like you aren't able to install the package at all etc (from what it looks like is gir1.2 outdated), the only solution seems to be using Ayatana's AppIndicator, sorry for not adding that in the pr at first

It also seems like it might be what this issue is about https://github.com/DevilXD/TwitchDropsMiner/issues/221

Maybe there's a better way to do it tbh, but it's the only solution i could fine that seems to have fixed it for me

Comment (guihkx) [reply build.spec:56]: It's strange that the regular AppIndicator doesn't work for you. Do you get any errors in the terminal, or something? What version of Raspberry Pi OS are you running?

>It also seems like it might be what this issue is about https://github.com/DevilXD/TwitchDropsMiner/issues/221

That's actually documentation issue, but [this comment](https://github.com/DevilXD/TwitchDropsMiner/issues/221#issuecomment-1591610740) explains what are the required dependencies.

File: .github/workflows/ci.yml
Comment (biast12) [reply .github/workflows/ci.yml:201]: From what it looks like does it not required more dependencies than what's already included, tho i'll have a bit deeper look just to be sure

Comment (guihkx) [reply .github/workflows/ci.yml:201]: Please do. Because I think at the very least you'll have to replace this dependency:

https://github.com/DevilXD/TwitchDropsMiner/blob/80f3e1b5f9002834b935ec3ca4927767bd1f8801/appimage/AppImageBuilder.yml#L51

By `gir1.2-ayatanaappindicator3-0.1`.

File: build.spec
Comment (biast12) [reply build.spec:56]: yes! 

When i try to run the `sudo apt install gir1.2-appindicator3-0.1 libcairo2-dev libgirepository1.0-dev pkg-config` command:
```sh
0.1 libcairo2-dev libgirepository1.0-dev pkg-config
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package gir1.2-appindicator3-0.1 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'gir1.2-appindicator3-0.1' has no installation candidate
```

When i try to run the build:
```sh
Running pyinstaller...
159 INFO: PyInstaller: 6.11.1, contrib hooks: 2024.10
159 INFO: Python: 3.11.2
160 INFO: Platform: Linux-6.6.51+rpt-rpi-2712-aarch64-with-glibc2.36
160 INFO: Python environment: /home/tobia/Desktop/TwitchDropsMiner/env
178 INFO: Module search paths (PYTHONPATH):
['/home/tobia/Desktop/TwitchDropsMiner',
 '/usr/lib/python311.zip',
 '/usr/lib/python3.11',
 '/usr/lib/python3.11/lib-dynload',
 '/home/tobia/Desktop/TwitchDropsMiner/env/lib/python3.11/site-packages',
 '/home/tobia/Desktop/TwitchDropsMiner']
520 INFO: Appending 'binaries' from .spec
521 INFO: Appending 'datas' from .spec
Unable to find '/usr/lib/aarch64-linux-gnu/girepository-1.0/AppIndicator3-0.1.typelib' when adding binary and data files.
PyInstaller build failed.
```
(this fails obviously because appindicator3 isn't installed)

With the issue did i refer to some of the later comments, sorry should have been more clear about that

Honestly i agree, the issue is just i can't find any other way other than installing it manually i guess

Did i bit more digging and yeah it looks like AppIndicator3 is deprecated which is why i can't install it, so it might be a good idea to move to AyatanaAppIndicator3 either way
https://github.com/Ulauncher/Ulauncher/issues/702#issuecomment-803027651
https://www.reddit.com/r/debian/comments/yqqa27/comment/ivpn05b/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

File: .github/workflows/ci.yml
Comment (biast12) [reply .github/workflows/ci.yml:201]: Yee that i already released, gonna do a push after a bit more research on this one exactly

File: build.spec
Comment (guihkx) [reply build.spec:56]: Oh, so AppIndicator-related packages got dropped in recent versions of Debian/Ubuntu... Gotcha. We should definitely switch to Ayatana then.

Comment (biast12) [reply build.spec:56]: Yeah at least that's what it looks like and that would explain why i can't install it too

Just to be 100% sure as you definitely know more than me on this, is there more than needs to be changed other than the thing in the workflow and in [AppImageBuilder.yml](https://github.com/DevilXD/TwitchDropsMiner/pull/617#discussion_r1842930395)? (they should be fixed now)

It works for me but idk if there's something missing within the actual scripts

File: appimage/AppImageBuilder.yml
Comment (guihkx) [inline appimage/AppImageBuilder.yml:56]: This is not incorrect, but I think it's best that we only include `gir1.2-ayatanaappindicator3-0.1` here, and then let AppImageBuilder automatically pick up the other dependencies. That way, in the unlikely scenario the list of dependencies change, we don't have take care of them manually.

```suggestion
      - gir1.2-ayatanaappindicator3-0.1
```

File: build.spec
Comment (guihkx) [inline build.spec:55]: I'm thinking this whole section can be simplified to something like this (untested):

```py
datas.append((Path(f"/usr/lib/{arch}-linux-gnu/girepository-1.0/AyatanaAppIndicator3-0.1.typelib"), "gi_typelibs"))
binaries.append((Path(f"/usr/lib/{arch}-linux-gnu/libayatana-appindicator3.so.1"), "."))
```

Because there's no real reason to artificially limit this to `x86_64` and `aarch64`, given that Ubuntu supports other architectures as well...

File: setup_env.sh
Comment (guihkx) [inline setup_env.sh:14]: In my opinion, a script like this should not be trying to install system packages. 

I understand the need of making things a bit user-friendlier, but this solution only really covers Debian-based distros that use `apt` to install packages. And even if we tried to support every distro out there, in my opinion we would be over-complicating things for very little gain.

Instead, we should document what system dependencies are necessary to run the script (later in the Wiki), and simply let the script fail when they're are not installed.

File: setup_env.sh
Comment (guihkx) [inline setup_env.sh:3]: It's considered "rude" for a script clear the terminal scrollback buffer like that.

```suggestion
```

File: build.sh
Comment (guihkx) [inline build.sh:3]: It's considered "rude" for a script clear the terminal scrollback buffer like that.

```suggestion
```

File: build.spec
Comment (biast12) [reply build.spec:55]: yeah that's a good idea i guess, the only issue i might see happening is some weird os not using that naming scheme, but i guess that's kinda a niche case compared to disallowing it, i'll pr this with the other review

File: setup_env.sh
Comment (biast12) [reply setup_env.sh:14]: yee didn't think about that, i'll remove it

File: setup_env.sh
Comment (biast12) [reply setup_env.sh:3]: 100% agree, i just "converted" the batch files which does cls, should i remove them from the batch files too?

File: build.sh
Comment (guihkx) [inline build.sh:11]: Is there a distro or file manager out there that lets us double-click `.sh` scripts to run them? I don't remember any that allows that by default...

If there's not, I think we could reduce friction and remove this and other calls to `read -p`.

I use KDE Plasma, and in Dolphin (their default file manager), they have this feature through the right-click menu (*Run In Konsole*), but even then, once the script finishes, they keep the terminal window open.

File: setup_env.sh
Comment (guihkx) [reply setup_env.sh:3]: I'd say so, but @DevilXD might think otherwise. :p

File: build.sh
Comment (guihkx) [inline build.sh:1]: Nitpick ("best practices"):

```suggestion
#!/usr/bin/env bash
```

File: setup_env.sh
Comment (guihkx) [inline setup_env.sh:1]: Nitpick ("best practices"):

```suggestion
#!/usr/bin/env bash
```

File: setup_env.sh
Comment (guihkx) [reply setup_env.sh:3]: I'd say so, but @DevilXD might think otherwise. :p

File: build.sh
Comment (biast12) [reply build.sh:11]: Nope, tho we might be able to make a workaround and just run the script directly? it would definitely make it much easier

```sh
# Check if the virtual environment exists
if [ ! -d "$dirpath/env" ]; then
    echo
    echo "No virtual environment found! Running setup_env.sh to set it up first."
    echo
    "$dirpath/setup_env.sh"
    if [ $? -ne 0 ]; then
        echo "Failed to set up the virtual environment."
        exit 1
    fi
fi
```

Comment (guihkx) [reply build.sh:11]: Sorry, my comment above this one was not clear.

I asked that because I assumed that the only reason you're using `read -p` to wait for user input, is in a scenario where users would run the script by double-clicking on them, and therefore those `read -p` calls would prevent the spawned terminal window from being closed upon script completion.

The Windows `.bat` scripts have this as well (through `pause` calls), to avoid command prompt window from being closed automatically too.

However, this is usually not done with Linux scripts, because most people will open an empty terminal window first, and then run the script later...

Comment (biast12) [reply build.sh:11]: ohh yeah sorry for misunderstanding, i do sometimes run the scripts by clicking on it and selecting "Execute in terminal", tho as you said most people open a clean terminal and then run it
image

Personally i think it's best to keep it, sure it's an extra step for those opening a clear terminal but i can definitely see it creating issues for new people opening them the other way

Comment (biast12) [reply build.sh:11]: Now the question is just, would it be better do keep it that way or add that it opens the script itself, personally do i think it would be nice, but i can also see why you would keep them separate

Comment (guihkx) [reply build.sh:11]: Yeah, it's fine to keep them, I just didn't know of a file manager that could do that. :p

Out of curiosity, if you choose "Execute in Terminal", will that file manager automatically close the spawned terminal window when the script finishes running?

Comment (biast12) [reply build.sh:11]: Yes! function wise does it work exactly like when you open a batch script by double clicking on it

Comment (guihkx) [reply build.sh:11]: >Now the question is just, would it be better do keep it that way or add that it opens the script itself, personally do i think it would be nice, but i can also see why you would keep them separate

I think it's better to keep these two scripts just for documentation purposes.

Since Windows also uses two scripts, we avoid having to rewrite parts of the Wiki.

File: build.spec
Comment (biast12) [reply build.spec]: Everything looks clear now? 👀

Comment (guihkx) [reply build.spec]: Linux-wise, yeah. I just want one minor nitpick.

File: build.spec
Comment (guihkx) [inline build.spec:53]: Please retain this comment

Also, can you please put these AppIndicator-related changes into a separate commit explaining the reason for the move to Ayatana's AppIndicator (i.e. the old one is deprecated), and finally `git rebase` your branch, melding some commits so the list of commits look cleaner? Thanks.

**Review by guihkx — APPROVED** [review]
Just one minor nitpick. I've only tested the Linux changes, and they  seem fine. Thanks for working on this!

File: build.spec
Comment (biast12) [reply build.spec:53]: Just to be sure i understand correctly, you want me to do a separate pr about the move to Ayatana's Appindicator?

Comment (guihkx) [reply build.spec:53]: It's not necessary to make a new pull request, you could use `git rebase` to fix the list of commits, and then force-push later. However, if you've never done that before, I think it's safer leave it like this.

Comment (biast12) [reply build.spec:53]: nonon, i know how to rebase, i'm thinking about the "can you please put these AppIndicator-related changes into a separate commit explaining the reason for the move to Ayatana's AppIndicator" part

Comment (guihkx) [reply build.spec:53]: That would be part of the rebase, too. Your commits would [look like this](https://github.com/DevilXD/TwitchDropsMiner/compare/master...guihkx:TwitchDropsMiner:example), for example.

Comment (biast12) [reply build.spec:53]: Sorry i misunderstood what you meant, should be all done now

**Comment by biast12** [comment]
ok nvm i fucked it up again, give me 2 😭

**Comment by biast12** [comment]
and done xD

**Comment by biast12** [comment]
Mostly only the packages that needs to be installed and mention that they can run the .sh files if they're on linux

@guihkx anything missing? i don't think there's more than that tbh

**Comment by guihkx** [comment]
I don't think anything else is missing. I've added Linux-specific instructions to the "[Setting up the environment, building and running](https://github.com/DevilXD/TwitchDropsMiner/wiki/Setting-up-the-environment,-building-and-running)" page in this diff (feel free to use/modify it to your liking):

```diff
diff --git a/Setting-up-the-environment,-building-and-running.md b/Setting-up-the-environment,-building-and-running.md
index dd4e586..11c5803 100644
--- a/Setting-up-the-environment,-building-and-running.md
+++ b/Setting-up-the-environment,-building-and-running.md
@@ -1,28 +1,64 @@
 # Prerequisites
 
-In order to setup the development environment for the application, you'll need to install both Python and Git. Here are two guides which will help you through the process:
+In order to setup the development environment for the application, you'll need to install both Python and Git.
+
+## Windows
+
+Here are two guides which will help you through the process:
 
 - [Git Installation](https://github.com/DevilXD/TwitchDropsMiner/wiki/Git-Installation)
 - [Python Installation](https://github.com/DevilXD/TwitchDropsMiner/wiki/Python-Installation)
 
 Continuing past this point, you should have both of them installed with the configuration and steps explained in each guide, and your computer should be restarted (see the "Final note" section in either of the two guides above for details). 
 
+## Linux
+
+This guide assumes you're on a modern version of a Ubuntu-based Linux distribution. If you're on a different distro, the name of packages used below will likely be different, and you will need to adapt them yourself.
+
+You'll need to install Git (`git`), Python 3 with Tkinter support (`python3-tk`) and Python's package installer (`python3-pip`).
+
+For system tray support, you'll also need Python 3 development headers (`python3-dev`), Cairo 2D development headers (`libcairo2-dev`), Pycairo development headers (`python3-cairo-dev`), GObject introspection data development headers (`libgirepository1.0-dev`), pkgconf (`pkg-config`) and a C compiler (`gcc`).
+
+In one `apt install` command:
+
+```sh
+sudo apt install git python3-tk python3-pip python3-dev libcairo2-dev python3-cairo-dev libgirepository1.0-dev pkg-config gcc
+```
+
 # Setup the environment
 
 To setup the runtime environment for the application, first, download the source code. If you're using `git`, you can `git clone https://github.com/DevilXD/TwitchDropsMiner` the project into a folder of choice. Otherwise, you can download the source code from the main project's page by using the button shown below, and then unzip it into a folder of choice.
 
 ![code-download](https://user-images.githubusercontent.com/4180725/206105115-5d94e594-9a79-4fda-b6b6-5ed487985208.png)
 
+## Windows
+
 Once you have the source code, run `setup_env.bat`. This will create an `env` folder next to the source code files, and install all required Python packages inside it. Watch out for permission errors, you may need to run it as Administrator.
 
+## Linux
+
+Once you have the source code, open a terminal window and use the `cd` command to navigate to the location of the source code. After that, run `./setup_env.sh`. This will create an `env` folder next to the source code files, and install all required Python packages inside it.
+
 # Running from source
 
+## Windows
+
 Run `run_dev.bat` to start the development build from source code. Starting without a console is perfectly fine, but you may want to start with a console if you'd expect errors to be printed out.
 
+## Linux
+
+First, run `source ./env/bin/activate` to activate the Python virtual environment. Then, run `python3 main.py` to start the application.
+
 # Building an executable
 
+## Windows
+
 If you'd want to build an executable, run `build.bat` - this will create a `dist` folder with the resulting executable inside. PyInstaller will be automatically installed into the environment (`env` folder) on the first build attempt, and thus might require running it as Administrator. Subsequent builds should complete just fine with normal permissions.
 
+## Linux
+
+If you'd want to build an executable, run `./build.sh` - this will create a `dist` folder with the resulting executable inside. PyInstaller will be automatically installed into the environment (`env` folder) on the first build attempt.
+
 # Manual instructions
 
 If you're one of those people who doesn't trust pre-made scripts, or you're interested in doing something those scripts haven't foreseen, here's some quick instructions which can get you up and running. Note that all commands assume you have Python and Git installed - if not, check out the prerequisites section at the top. Also, it's assumed you'll start your command prompt in the source code's directory, setting it as the current working directory.
@@ -32,4 +68,4 @@ If you're one of those people who doesn't trust pre-made scripts, or you're inte
 - With the venv activated, running `pip install -r requirements.txt` should install all required packages into the virtual environment. Please ensure it's activated properly before doing so, otherwise the packages will be installed in your system-global Python installation instead, where you won't be able to use them. A warning stating a pip update can be performed is normal and a non-mandatory thing for you to take care of.  
 - With the packages installed and still in the activated venv, you can run the application with `python main.py`. This will start it with an extra console window, to which any eventual Python error might be printed into. Because during normal operation it's entirely not needed, you can hide it by starting the application with `pythonw main.py` instead.
 - If you'd wish not to have to activate the venv every time you run the application, it's also possible to use the venv's executables directly. `./env/Scripts/python main.py` (`bin` replaces `Scripts` on Linux) will start the application from the venv, without having to activate it first. Feel free to create yourself a shortcut that'll do that for you, or consider building the application into an executable form.  
-- In order to build an executable, you'll need an extra package called PyInstaller. Install it into **an activated venv** via `pip install pyinstaller`, then with the venv still activated, run `pyinstaller build.spec`. All information related to building the application are already defined in the `build.spec` file for your convenience. Both Windows and Linux executables are supported.
\ No newline at end of file
+- In order to build an executable, you'll need an extra package called PyInstaller. Install it into **an activated venv** via `pip install pyinstaller`, then with the venv still activated, run `pyinstaller build.spec`. All information related to building the application are already defined in the `build.spec` file for your convenience. Both Windows and Linux executables are supported.

```

**Comment by guihkx** [comment]
I just noticed `setup_env.sh` and `build.sh` don't have execute permissions.

@biast12 Can you please do that?

**Comment by biast12** [comment]
I can change it but it for some reason won't allow me to commit it :/

**Comment by guihkx** [comment]
That's odd... Do you get an error or something?

**Comment by biast12** [comment]
Nope, it just doesn't change anything binary wise about the files so github can't see any difference, i just tried to force it but to no avail

**Comment by guihkx** [comment]
If you want, feel free to [add me as a collaborator](https://github.com/biast12/TwitchDropsMiner/settings/access) to your TwitchDropsMiner fork (so I'm given commit access), and I can try to fix it for you.

But if you think that's too much work or not worth it, I can do it later in a future PR. And you can revert your last empty commit as well.

**Comment by biast12** [comment]
I just added you, pretty sure that's the best and easiest route for us all xD

**Comment by guihkx** [comment]
Should be good now. I've removed myself as a collaborator too. :p

**Comment by biast12** [comment]
Hey @DevilXD, the pr is done / ready to merge if you got time to look at it at some point

**Comment by DevilXD** [comment]
I understand. However, I haven't been able to find enough time to be able to look at this in detail, plus the wiki documentation has to be altered to include the Linux instructions. I'm currently back in work delegation for the rest of this year, so this will need to wait until about the middle of January. I'll try to get to it on my off time soft-scheduled for 27th-Dec to 3rd-Jan, but all stars are currently pointing at the off time not being a thing, and me working through out this time as well.

**Comment by biast12** [comment]
Yep all good, just wanted to be sure you hadn't missed it, hope you can get some free time (not just for this but in general also), and I wish you a merry christmas and a happy new year for the time being.

File: setup_env.sh
Comment (DevilXD) [inline setup_env.sh:51]:
```suggestion
read -p "Press any key to continue..."

```

Again, one more style-related empty new line at the end.

File: build.sh
Comment (DevilXD) [inline build.sh:36]:
```suggestion
read -p "Press any key to continue..."

```

Again, as a style thing, the file should end with a single empty new line.

**Review by DevilXD — CHANGES_REQUESTED** [review]
I've finally found some free time for this, and already finished doing the wiki page changes. Thank you @guihkx for the diff 🙂

I've also reviewed the changes of this PR, and I have some remarks, which you can find below. Most of them are trivial, but there's two places that would need to be changed for me to approve this. All the details of them can be found in the discussions below.

File: setup_env.bat
Comment (biast12) [reply setup_env.bat:38]: would it be best to just remove the whole exit line for both? so it still echos that it fails but doesn't exit

**Comment by biast12** [comment]
Got all the changes ready to commit, do you want me to just do a commit or rebase again?

File: setup_env.bat
Comment (DevilXD) [reply setup_env.bat:38]: Honestly: I don't really know. Either of the solutions is okay by me. The only thing I can see happening here, is that if the installation fails, and if there's anything fail-related displayed to the user, they may get confused and try to report it, which as explained, isn't really that big of a deal. Also, the first installation failing would mean the `wheel` installation, as well as the rest of the requirements, will also most likely fail, but that's something we can't really predict or guarantee would happen, so...

If you're unsure, due to the above, it'd probably be best not to print anything fail-related to the console.

**Comment by DevilXD** [comment]
Up to you, really. Just so you know, during merging, I rebase changes where the commits history is kept clean, otherwise if the commits are just "Updated twitch.py" (commit messages aren't descriptive), I do a squash and give the merge commit a more descriptive message (usually for translations PRs).

Since the commits in this PR are looking great so far, I'll be doing a rebase, so rebasing is preferred to keep everything clean. A commit is fine too, if you don't want to bother though =)

File: setup_env.bat
Comment (biast12) [reply setup_env.bat:38]: Oh do you also want me to remove it from the requirements part? since that would fail too

Comment (DevilXD) [reply setup_env.bat:38]: The `requirements.txt` failing to install is an actual problem, so it can stay there. The failure would be obvious from the command output already though, so... again, not entirely sure. Up to you as well.

Comment (biast12) [reply setup_env.bat:38]: Perfect, then i'll just let it be, everything should be ready then

**Comment by biast12** [comment]
And that should be it, i just did the commit as it's probably better for all of us 😭

File: build.sh
Comment (DevilXD) [inline build.sh:20]:
```suggestion
        echo
        echo "Failed to install PyInstaller."
```

**Review by DevilXD — COMMENTED** [review]
I've taken another look, and I've noticed a couple more things to resolve. Most of them are straight forward consistency changes between the two scripts, like for ex. the shell script including an echo description, while the batch script didn't.

An important thing about the batch script: On Windows, the scripts are usually ran standalone, via double-click. This opens and attaches a separate console window to the script, that closes together with the script finishing at the `exit` command. Because of that, to be able to actually "see" any output being printed out for more than a fraction of a second before the console window disappears, the script has to incorporate a `pause` command, that pauses the execution until any key is pressed. Without that, those "failed to install ..." lines are pretty much never going to be seen, and the window will just "flash" (open and almost immediately close) without giving the user a change to read anything. The extra echo commands are there to space the messages out a little, as that improves readability.

Fun fact, the `echo:` command isn't a real command, it's a hack that makes `echo` print nothing, which serves as just printing a new, empty line. Using just `echo` prints the current echo status. Any non-special character just after `echo` (but without a space) works: 
picture

**Comment by biast12** [comment]
And there, i also changed the "All done!" echo to something a bit more fitting to the consistency of the scripts / "Environment setup completed successfully.", everything should be fine now tho

**Review by DevilXD — APPROVED** [review]
Everything looks good now =)

**Pull request merged**
